Batteries/Tactic/PrintOpaques.lean

```diff
@@ -0,0 +1,76 @@
+/-
+Copyright (c) 2022 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro
+-/
+import Lean.Elab.Command
+import Lean.Util.FoldConsts
+
+open Lean Elab Command
+
+namespace Batteries.Tactic.CollectOpaques
+
+/-- Collects the result of a `CollectOpaques` query. -/
+structure State where
+  /-- The set visited constants. -/
+  visited : NameSet := {}
+  /-- The collected opaque defs. -/
+  opaques : Array Name := #[]
+
+/-- The monad used by `CollectOpaques`. -/
+abbrev M := ReaderT Environment <| StateT State MetaM
+
+/-- Collect the results for a given constant. -/
+partial def collect (c : Name) : M Unit := do
+  let collectExpr (e : Expr) : M Unit := e.getUsedConstants.forM collect
+  let s ← get
+  unless s.visited.contains c do
+    modify fun s => { s with visited := s.visited.insert c }
+    let env ← read
+    match env.find? c with
+    | some (ConstantInfo.ctorInfo _)
+    | some (ConstantInfo.recInfo _)
+    | some (ConstantInfo.inductInfo _)
+    | some (ConstantInfo.quotInfo _)   =>
+      pure ()
+    | some (ConstantInfo.defnInfo v)
+    | some (ConstantInfo.thmInfo v)    =>
+      unless ← Meta.isProp v.type do collectExpr v.value
+    | some (ConstantInfo.axiomInfo v)
+    | some (ConstantInfo.opaqueInfo v) =>
+      unless ← Meta.isProp v.type do
+        modify fun s => { s with opaques := s.opaques.push c }
+    | none                             =>
+      modify fun s => { s with opaques := s.opaques.push c }
```

Comment:
That case doesn't occur in normal usage. It can occur if `collect` is called directly with an invalid name. It's unclear what is the right thing to do. `collectAxioms` just ignores this case, I suppose Mario thought that an unknown constant is opaque. I'm changing it to panic, which is the safest thing to do.
```suggestion
      panic! s!"unknown constant {c}"
```

Reply:
maybe panic and then push it to the opaques list?

Reply:
Right now the panic returns `pure ()`, which basically ignores this case the same way `collectAxioms` does. Is there a reason you think it's better to add it to the opaques list? It seems to me that would be error prone since users might expect that all the names returned correspond to actual constants.

Reply:
This command is used like `#print axioms`, it is used to detect dependencies outside an expected set. So if something weird is going on, the conservative position is to make sure the situation is reported, and not ignore it.